daemon/pipeline.c

```diff
@@ -0,0 +1,372 @@
+
+#include <glib.h>
+#include <stdbool.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <gst/gst.h>
+
+#include <common.h>
+#include <modules.h>
+#include <gdbus-util.h>
+#include <log.h>
+
+#include "dbus-interface.h"
+
+static MachinelearningServicePipeline *g_gdbus_instance = NULL;
+
+static int probe_pipeline (void *data);
+static void init_pipeline (void *data);
+static void exit_pipeline (void *data);
+
+static GHashTable *pipeline_table;
+
+typedef struct _pipeline {
+  GstElement *element;
+  gint64 id;
+  GMutex lock;
+  gchar *description;
+  // Is this enough?
+  // Is register any gst related callback valid?
+  // When there occur error or state change -> should broadcast it via d-bus to clients
+} pipeline_s;
+
+// See ml_pipeline_destroy
+static void _pipeline_free (gpointer data)
+{
+  pipeline_s *p;
+
+  if (!data) {
+    g_critical ("internal error, the data should not be NULL");
+    return;
+  }
+
+  p = (pipeline_s *) data;
+  g_warning ("delete pipeline whose ID is %" G_GINT64_FORMAT, p->id);
+
+  // TODO should stop and broadcast the info to D-BUS
+  // gst_element_get_state (p->element, ...);
+  g_mutex_lock (&p->lock);
+  if (p->element)
+    gst_object_unref (p->element);
+  g_mutex_unlock (&p->lock);
+
+  g_free (p->description);
+  g_mutex_clear (&p->lock);
+
+  g_free (p);
+}
+
+static int _check_pipeline_description (const gchar *description)
+{
+  // TODO implement the logic for description checking
+  g_warning ("check whether valid or not of %s", description);
+  return 0;
+}
+
+static gboolean dbus_cb_core_register_pipeline (MachinelearningServicePipeline *obj,
+        GDBusMethodInvocation *invoc, const gchar *description, gpointer user_data)
+{
+  int ret = 0;
+  GError *err = NULL;
+  GstElement *pipeline = NULL;
+  pipeline_s *p;
+
+  g_warning ("Receive description: %s", description);
+
+  // check pipeline
+  // 1. it should have structure of  tensor_query_serversrc ! ... ! tensor_query_serversink
+  // 2. It should not contain tensor_sink, appsink, appsrc, ...
+  ret = _check_pipeline_description (description);
+  if (ret) {
+    g_critical ("Invalid pipeline description It should start and end with tensor_query_server* and not contain appsrc/appsink/tensor_sink");
+  }
+
+  // launch
+  pipeline = gst_parse_launch (description, &err);
+  if (!pipeline || err) {
+    g_critical ("gst_parse_launch with %s Failed. error msg: %s", description, (err) ? err->message : "unknown reason");
+    g_clear_error (&err);
+
+    if (pipeline)
+      gst_object_unref (pipeline);
+
+    machinelearning_service_pipeline_complete_register_pipeline (obj, invoc, -1, -1);
+    return FALSE;
+  }
+
+  p = g_new0 (pipeline_s, 1);
+  p->element = pipeline;
+  p->id = g_get_monotonic_time ();
+  p->description = g_strdup (description);
+  g_mutex_init (&p->lock);
+
+  g_hash_table_insert (pipeline_table, GINT_TO_POINTER (p->id), p);
```

Comment:
Be careful. g_hash_table is not thread safe.

If the functions in this .c file can be invoked by DBUS calls, you need to protect the global variable, ```pipeline_table```, ready for manipulations by other processes.